File: Algorithms_and_Data_structures/ADS_2021/w4_BST/lab4/leetcode_additional/1_108.py
# 108. Convert Sorted Array to Binary Search Tree


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def sortedArrayToBST(self, nums: list[int]):
        
        def recursion(left, right):
            if (left > right):
                return None
            mid = (left + right) // 2
            node = TreeNode(nums[mid])
            node.left = recursion(left, mid - 1)
            node.right = recursion(mid + 1, right)
            return node 
        
        return recursion(0, len(nums) - 1)

File: Algorithms_and_Data_structures/ADS_2021/w4_BST/lab4/leetcode_additional/1_108.cpp
// 108. Convert Sorted Array to Binary Search Tree


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        int sz = nums.size();
        if (sz == 0){
            return NULL;
        }
        return recursion(nums, 0, sz - 1);
    }
    
    TreeNode* recursion(vector<int>& v, int left, int right){
        if (left > right){
            return NULL;
        }
        int middle = (left + right) / 2;

        TreeNode* node; 
        node = new TreeNode(v[middle]);
        node->left = recursion(v, left, middle - 1);
        node->right = recursion(v, middle + 1, right);
        return node;
    }
};